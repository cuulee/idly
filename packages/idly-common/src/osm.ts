export type Entity = Node | Way | Relation;
export type Entities = Set<Node | Way | Relation>;
export type EntityId = string;
export type NodeId = string;
export type WayId = string;

export type ParentWays = Map<NodeId, Set<WayId>>;
// Table used to map id -> entity
export type EntityTable = Map<EntityId, Entity>;

export type Tags = Map<string, string>;

export type FeaturePropsTable = Map<EntityId, FeatureProps>;
export type FeatureProps = { [key: string]: string };

export type NodeGeometry =
  | OsmGeometry.POINT
  | OsmGeometry.VERTEX
  | OsmGeometry.VERTEX_SHARED;

export enum OsmGeometry {
  POINT = 'point',
  VERTEX = 'vertex',
  VERTEX_SHARED = 'vertex_shared', // the one shared among multiple ways
  AREA = 'area',
  LINE = 'line',
  RELATION = 'relation'
}

export enum EntityType {
  NODE = 'node',
  WAY = 'way',
  RELATION = 'relation'
}

export interface Node {
  readonly id: EntityId;
  readonly tags: Tags;
  readonly type: EntityType.NODE;
  readonly loc: LngLat;
  readonly attributes: Attributes;
}

export interface Way {
  readonly id: EntityId;
  readonly type: EntityType.WAY;
  readonly tags: Tags;
  readonly attributes: Attributes;
  readonly nodes: EntityId[];
}

export interface Relation {
  readonly id: EntityId;
  readonly type: EntityType.RELATION;
  readonly tags: Tags;
  readonly members: RelationMember[];
  readonly attributes: Attributes;
}

export interface Attributes {
  readonly visible?: boolean;
  readonly version?: number;
  readonly timestamp?: string;
  readonly changeset?: string;
  readonly uid?: string;
  readonly user?: string;
}

export interface RelationMember {
  readonly id: string;
  readonly type: string;
  readonly role: string;
}

export interface LngLat {
  readonly lat: number;
  readonly lon: number;
}
