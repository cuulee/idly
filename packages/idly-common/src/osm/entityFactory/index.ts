export { attributesFactory } from './attributesFactory';
export { nodeFactory } from './nodeFactory';
export { relationFactory } from './relationFactory';
export { tagsFactory } from './tagsFactory';
export { wayFactory } from './wayFactory';
