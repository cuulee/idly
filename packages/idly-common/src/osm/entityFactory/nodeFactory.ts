import { lngLatFactory } from '../../geo/lngLatFactory';
import { deepFreeze } from '../../misc/deepFreeze';

import {
  Attributes,
  EntityId,
  EntityType,
  LngLat,
  Node,
  Tags,
} from '../structures';
import { attributesFactory } from './attributesFactory';
import { tagsFactory } from './tagsFactory';

export function nodeFactory(
  {
    id,
    tags = tagsFactory(),
    loc = lngLatFactory([0, 0]),
    attributes = attributesFactory(),
  }: {
    id: EntityId;
    tags?: Tags;
    loc?: LngLat;
    attributes?: Attributes;
  },
  freeze = true
): Node {
  return deepFreeze<Node>(
    {
      attributes: attributesFactory(attributes),
      id,
      loc,
      tags: tagsFactory(tags),
      type: EntityType.NODE,
    },
    freeze
  );
}
