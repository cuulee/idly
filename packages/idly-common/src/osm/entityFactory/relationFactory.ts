import { deepFreeze } from '../../misc/deepFreeze';
import {
  Attributes,
  EntityId,
  EntityType,
  Relation,
  RelationMember,
  Tags,
} from '../structures';
import { attributesFactory } from './attributesFactory';
import { tagsFactory } from './tagsFactory';

export function relationFactory(
  {
    id,
    tags = tagsFactory(),
    attributes = attributesFactory(),
    members = [],
  }: {
    id: EntityId;
    tags?: Tags;
    attributes?: Attributes;
    members?: RelationMember[] | ReadonlyArray<RelationMember>;
  },
  freeze = true
): Relation {
  return deepFreeze<Relation>(
    {
      attributes: attributesFactory(attributes),
      id,
      members,
      tags: tagsFactory(tags),
      type: EntityType.RELATION,
    },
    freeze
  );
}
