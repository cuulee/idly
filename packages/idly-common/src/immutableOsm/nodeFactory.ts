import { $attributesGen } from './attributesGen';
import { $genLngLat } from './genLngLat';
import { $tagsFactory } from './tagsFactory';
import { $Node, $__Node } from '../immutableOsm';
import { genLngLat } from '../osm/genLngLat';
import { attributesGen } from '../osm/attributesGen';
import { Attributes, EntityId, EntityType, LngLat, Node, Tags } from '../osm';
import { recordify } from 'typed-immutable-record';

export function nodeFactory(n: Node): $Node {
  var $tags = $tagsFactory(n.tags);
  var $loc = $genLngLat(n.loc);
  var $attr = $attributesGen(n.attributes);
  return recordify<$__Node, $Node>({
    id: n.id,
    tags: $tags,
    type: EntityType.NODE,
    loc: $loc,
    attributes: $attr
  });
}
