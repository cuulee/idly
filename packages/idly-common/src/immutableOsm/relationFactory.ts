import { List as $List } from 'immutable';
import { $relationMemberGen } from './relationMemberGen';
import { $attributesGen } from './attributesGen';
import { $tagsFactory } from './tagsFactory';
import { $Relation, $__Relation } from '../immutableOsm';
import { attributesGen } from '../osm/attributesGen';
import {
  Attributes,
  EntityId,
  EntityType,
  Relation,
  RelationMember,
  Tags
} from '../osm';
import { recordify } from 'typed-immutable-record';

export function $relationFactory(r: Relation): $Relation {
  var $tags = $tagsFactory(r.tags);
  var $attr = $attributesGen(r.attributes);
  var $members = $List(r.members.map($relationMemberGen));

  return recordify<$__Relation, $Relation>({
    id: r.id,
    type: EntityType.RELATION,
    tags: $tags,
    attributes: $attr,
    members: $members
  });
}
