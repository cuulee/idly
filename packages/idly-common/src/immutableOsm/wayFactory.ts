import { $attributesGen } from './attributesGen';
import { $tagsFactory } from './tagsFactory';
import { $Way, $__Way } from '../immutableOsm';
import { Attributes, EntityId, EntityType, Tags, Way } from '../osm';
import { attributesGen } from '../osm/attributesGen';
import { recordify } from 'typed-immutable-record';
import { List as $List } from 'immutable';

export function $wayFactory(way: Way): $Way {
  var $tags = $tagsFactory(way.tags);
  var $attr = $attributesGen(way.attributes);
  var $nodes = $List(way.nodes);
  return recordify<$__Way, $Way>({
    id: way.id,
    type: EntityType.WAY,
    tags: $tags,
    attributes: $attr,
    nodes: $nodes
  });
}
