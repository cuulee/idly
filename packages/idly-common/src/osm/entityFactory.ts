import { nodeFactory } from '../osm/nodeFactory';
import { relationFactory } from '../osm/relationFactory';
import { EntityType, Tags } from '../osm/structures';
import { Entity, Node, Relation, Way } from './structures';
import { tagsFactory } from './tagsFactory';
import { wayFactory } from './wayFactory';

export function entityFactory(entity: any): Entity {
  if (!entity.id) {
    throw new Error('not a valid entity');
  }

  if (typeof entity.tags === 'string') {
    entity.tags = tagsFactory(JSON.parse(entity.tags));
  } else if (Array.isArray(entity.tags)) {
    entity.tags = tagsFactory(entity.tags);
  } else if (!entity.hasOwnProperty('tags')) {
    entity.tags = tagsFactory([]);
  }

  if (!(entity.tags instanceof Tags)) {
    throw new Error('not proper tags');
  }

  switch (entity.type) {
    case EntityType.NODE:
      return nodeFactory(entity);

    case EntityType.WAY:
      return wayFactory(entity);

    case EntityType.RELATION:
      return relationFactory(entity);

    default:
      throw new Error('not a valid entity');
  }
}
