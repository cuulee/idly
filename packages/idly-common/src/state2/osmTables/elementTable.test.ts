import { nodeFactory } from '../../osm/entityFactory/nodeFactory';
import { relationFactory } from '../../osm/entityFactory/relationFactory';
import { wayFactory } from '../../osm/entityFactory/wayFactory';
import { Entity } from '../../osm/structures';
import { setCreate } from '../helper';
import {
  Element,
  elementTableBulkAdd,
  elementTableCreate,
} from './elementTable';

const n1 = nodeFactory({
  id: 'n1',
});
const n2 = nodeFactory({
  id: 'n2',
});
const n3 = nodeFactory({
  id: 'n3',
});
const n4 = nodeFactory({
  id: 'n4',
});
const w1 = wayFactory({
  id: 'w1',
  nodes: ['n1', 'n2'],
});

const w2 = wayFactory({
  id: 'w2',
  nodes: ['n1', 'n3'],
});

const r1 = relationFactory({
  id: 'r1',
  members: [{ id: 'n4', ref: 'n4' }, { id: 'w2', ref: 'w2' }],
});

const dummyElement = (
  entity: Entity,
  parentWays = setCreate<string>(),
  parentRelations = setCreate<string>()
): Element => ({
  entity,
  parentRelations,
  parentWays,
});

it('adds entities', () => {
  const table = elementTableCreate();
  elementTableBulkAdd(table, [n1, n2]);
  expect(table).toEqual(
    new Map([['n1', dummyElement(n1)], ['n2', dummyElement(n2)]])
  );
});
it('adds parentways', () => {
  const table = elementTableCreate();
  elementTableBulkAdd(table, [n1, n2]);
  elementTableBulkAdd(table, [w1]);
  expect(table).toEqual(
    new Map([
      ['n1', dummyElement(n1, setCreate(['w1']))],
      ['n2', dummyElement(n2, setCreate(['w1']))],
      ['w1', dummyElement(w1)],
    ])
  );
});

it('adds everything correctly', () => {
  const t1 = elementTableCreate();
  elementTableBulkAdd(t1, [n1, n2, w2, w1, n3, r1, n4]);

  expect(t1).toMatchSnapshot();
});

it('adds everything correctly', () => {
  const t1 = elementTableCreate();
  elementTableBulkAdd(t1, [n1, n2, w2, w1, n3, r1, n4]);

  expect(t1.get('n1')).toMatchSnapshot();
});

it('throw error when entity node found in table', () => {
  const t1 = elementTableCreate();
  expect(() => elementTableBulkAdd(t1, [n1, w2])).toThrowError();
});
