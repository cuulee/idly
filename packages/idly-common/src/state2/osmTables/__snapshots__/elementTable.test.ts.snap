// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`adds everything correctly 1`] = `
Map {
  "n1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n1",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w2",
      "w1",
    },
  },
  "n2" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n2",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w1",
    },
  },
  "w2" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w2",
      "nodes": Array [
        "n1",
        "n3",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {
      "r1",
    },
    "parentWays": Set {},
  },
  "w1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "w1",
      "nodes": Array [
        "n1",
        "n2",
      ],
      "tags": Object {},
      "type": "way",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n3" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n3",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {},
    "parentWays": Set {
      "w2",
    },
  },
  "r1" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "r1",
      "members": Array [
        Object {
          "id": "n4",
          "ref": "n4",
        },
        Object {
          "id": "w2",
          "ref": "w2",
        },
      ],
      "tags": Object {},
      "type": "relation",
    },
    "parentRelations": Set {},
    "parentWays": Set {},
  },
  "n4" => Object {
    "entity": Object {
      "attributes": Object {
        "changeset": undefined,
        "timestamp": undefined,
        "uid": undefined,
        "user": undefined,
        "version": undefined,
        "visible": undefined,
      },
      "id": "n4",
      "loc": Object {
        "lat": 0,
        "lon": 0,
      },
      "tags": Object {},
      "type": "node",
    },
    "parentRelations": Set {
      "r1",
    },
    "parentWays": Set {},
  },
}
`;

exports[`adds everything correctly 2`] = `
Object {
  "entity": Object {
    "attributes": Object {
      "changeset": undefined,
      "timestamp": undefined,
      "uid": undefined,
      "user": undefined,
      "version": undefined,
      "visible": undefined,
    },
    "id": "n1",
    "loc": Object {
      "lat": 0,
      "lon": 0,
    },
    "tags": Object {},
    "type": "node",
  },
  "parentRelations": Set {},
  "parentWays": Set {
    "w2",
    "w1",
  },
}
`;
